### PR TITLE
Feature/container install

### PR DESCRIPTION
**Pull request by jemten**
### This PR adds | fixes:

- Cache images instead of downloading sif.
- Remove proxy bin

### How to test:

- Automatic and continuous test suite

### Expected outcome:
- [ ] Installation, unit and integration test suite pass

### Review:
- [ ] Code review
- [ ] Tests pass

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes. If applicable record manual test results in PR header
- [ ] **MINOR** - when you add functionality in a backwards compatible manner. If applicable record manual test results in PR header
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: lib/MIP/Environment/Container.pm [hunk 1]
```diff
@@ -429,12 +354,12 @@ sub run_container {
 
 sub set_executable_container_cmd {
 
-## Function : Set executable command depending on container manager
-## Returns  :
-## Arguments: $active_parameter_href => The active parameters for this analysis hash {REF}
-##          : $container_href        => Containers hash {REF}
-##          : $container_manager     => Container manager
-##          : $parameter_href        => Parameter hash {REF}
+    ## Function : Set executable command depending on container manager
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:357]: Align header

File: lib/MIP/Environment/Container.pm [hunk 2]
```diff
@@ -476,20 +401,68 @@ sub set_executable_container_cmd {
...
+            container_href                   => $container_href,
+            container_manager                => $container_manager,
+            recipe_executable_bind_path_href => \%recipe_executable_bind_path,
+        }
+    );
+
+    return %container_cmd;
+}
+
+sub build_container_cmd {
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:422]: Order subs AB

File: lib/MIP/Program/Conda.pm [hunk 3]
```diff
@@ -26,7 +26,7 @@ BEGIN {
     our $VERSION = 1.19;
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Conda.pm:26]:
```suggestion
    our $VERSION = 1.20;
```

File: lib/MIP/Program/Singularity.pm [hunk 4]
```diff
@@ -38,7 +38,7 @@ sub singularity_exec {
 ## Arguments: $bind_paths_ref                 => Array with paths to bind {REF}
 ##          : $filehandle                     => Filehandle to write to
 ##          : $image                          => Singularity container name
-##          : $singularity_container_cmds_ref => Array with commands to be executed inside container {REF}
+##          : $container_cmds_ref => Array with commands to be executed inside container {REF}
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Singularity.pm:41]: Very good that we standardise! Align fat comma

File: lib/MIP/Recipes/Install/Mip_scripts.pm [hunk 5]
```diff
@@ -39,76 +39,39 @@ sub install_mip_scripts {
...
-
     };
 
     check( $tmpl, $arg_href, 1 ) or croak q{Could not parse arguments!};
 
     use MIP::Check::Installation qw{ check_mip_executable };
 
-    ## Unpack parameters
-    my $mip_scripts_version = $mip_scripts_parameters_href->{version};
+    Readonly my $FIVE => 5;
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Install/Mip_scripts.pm:63]:
```suggestion
    Readonly my $MOVE_DIRS_UP => 5;
```

File: lib/MIP/Recipes/Install/Mip_scripts.pm [hunk 6]
```diff
@@ -39,76 +39,39 @@ sub install_mip_scripts {
...
-            quiet    => $quiet,
-            verbose  => $verbose,
-        }
-    );
+    my $log = Log::Log4perl->get_logger($LOG_NAME);
 
-    ## Store original working directory
-    my $pwd = cwd();
+    ## Get mip directory relative to this file since $Bin might have been set already
+    my $mip_dir_path = path(__FILE__)->parent($FIVE);
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Install/Mip_scripts.pm:74]:
```suggestion
    my $mip_dir_path = path(__FILE__)->parent($MOVE_DIRS_UP);
```

File: lib/MIP/Recipes/Install/Mip_scripts.pm [hunk 7]
```diff
@@ -140,79 +103,57 @@ sub install_mip_scripts {
...
-                file_path  => $file_path,
-                filehandle => $filehandle,
-                permission => q{a+x},
-            }
-        );
-        say {$filehandle} $NEWLINE;
+        my $src_path = catfile( $mip_dir_path,      $script );
+        my $dst_path = catfile( $conda_prefix_path, q{bin}, $script );
+        path($src_path)->copy($dst_path);
+        path($dst_path)->chmod(q{a+x});
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Install/Mip_scripts.pm:156]: Neat!

File: lib/MIP/Script/Setup_script.pm [hunk 8]
```diff
@@ -827,8 +644,8 @@ sub setup_script {
     # Build bash shebang line
     build_shebang(
         {
-            bash_bin_path => catfile( dirname( dirname( devnull() ) ), qw{ bin bash } ),
-            filehandle    => $filehandle,
+            bash_bin_path      => catfile( rootdir(), qw{ bin bash } ),
```
Comment (henrikstranneheim) [inline lib/MIP/Script/Setup_script.pm:647]: 👍

File: lib/MIP/Test/Fixtures.pm [hunk 9]
```diff
@@ -453,4 +452,64 @@ sub test_standard_cli {
     return $verbose;
 }
 
+sub test_constants {
+
+    ## Function : Generate standard command line interface for test scripts
```
Comment (henrikstranneheim) [inline lib/MIP/Test/Fixtures.pm:457]: Align headers

File: lib/MIP/Test/Fixtures.pm [hunk 10]
```diff
@@ -453,4 +452,64 @@ sub test_standard_cli {
...
+
+    use MIP::Constants qw{ set_test_constants };
+
+    if ( not keys %{$test_process_return_href} ) {
+
+        %{$test_process_return_href} = (
+            buffers_ref        => [],
+            error_message => undef,
+            stderrs_ref        => [],
+            stdouts_ref        => [qw{ 1 }],
```
Comment (henrikstranneheim) [inline lib/MIP/Test/Fixtures.pm:500]:
```suggestion
            stdouts_ref        => [qw{ one }],
```
or
            stdouts_ref        => [( 1, )],

File: t/data/test_data/reference_info.yaml [hunk 11]
```diff
@@ -1,4 +1,4 @@
 ---
 mip:
-  path: /path/MIP/t/data/modules/miniconda/envs/mip_ci/bin/mip
+  path: /Users/andersjemt/aj/MIP/t/data/modules/miniconda/envs/mip_ci/bin/mip
```
Comment (henrikstranneheim) [inline t/data/test_data/reference_info.yaml:3]:
```suggestion
  path: /path/MIP/t/data/modules/miniconda/envs/mip_ci/bin/mip
```

File: t/get_conda_path.t [hunk 12]
```diff
@@ -58,17 +59,20 @@ diag(   q{Test get_conda_path from Parameter.pm v}
...
-$is_not_ok = get_conda_path( { bin_file => q{not_a_conda_bin} } );
-
-## Then return false
-is( $is_not_ok, undef, q{Did not find conda path} );
+trap {
+    get_conda_path( { bin_file => q{not_a_conda_bin} } )
+};
+## Then print fatal log message and exit
+ok( $trap->exit, q{Exit when no conda is found } );
+like( $trap->stderr, qr/FATAL/xms, q{Throw fatal log message} );
```
Comment (henrikstranneheim) [inline t/get_conda_path.t:76]: Is there a specific error message to catch? FATAL is a bit to general

**Review by henrikstranneheim — APPROVED** [review]
Fantastic 🥇  Anders! This simplifies a lot and brings such nice new functionality

**Comment by jemten** [comment]
any thoughts @ramprasadn? Otherwise I'll go ahead and merge

**Review by ramprasadn — APPROVED** [review]
Nice! Its always good to see so many lines of code go away :D  👍

**Pull request merged**